test/src/q_test.cpp

```diff
@@ -0,0 +1,229 @@
+#include "bm_os.h"
+#include "fff.h"
+#include "gtest/gtest.h"
+#include <helpers.hpp>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+DEFINE_FFF_GLOBALS
+
+#include "q.h"
+
+class queue_test : public ::testing::Test {
+protected:
+  rnd_gen RND;
+  static constexpr uint32_t queue_buf_size = 4096;
+
+  queue_test() {}
+
+  ~queue_test() override {}
+
+  void SetUp() override {}
+
+  void TearDown() override {}
+
+  void enqueue_dequeue_check(Q *queue) {
+    // Ensure there is enough space for items in the queue, make queue items small
+    constexpr size_t num_queue_items = 10;
+    constexpr size_t queue_item_size = (queue_buf_size / num_queue_items) -
+                                       (queue_buf_size / (num_queue_items * 2));
+    uint8_t queue_item_bufs[num_queue_items][queue_item_size] = {};
+
+    for (size_t i = 0; i < num_queue_items; i++) {
+      RND.rnd_array(queue_item_bufs[i], queue_item_size);
+      EXPECT_EQ(q_enqueue(queue, queue_item_bufs[i], queue_item_size), BmOK);
+    }
+
+    // Ensure that failure occurs if queue item is larger than space available
+    constexpr size_t queue_big_buf_size =
+        queue_buf_size - (num_queue_items * queue_item_size);
+    uint8_t queue_big_buf[queue_big_buf_size] = {};
+    RND.rnd_array(queue_big_buf, queue_big_buf_size);
+    EXPECT_NE(q_enqueue(queue, queue_big_buf, queue_big_buf_size), BmOK);
+
+    // Dequeue buffer and ensure it equals what was enqueued
+    uint8_t comp_buf[queue_item_size] = {0};
+
+    for (size_t i = 0; i < num_queue_items - 1; i++) {
```

Comment:
It seems like this "minus one" is a typo or leftover from earlier work. Tests still pass without it. Remove it?

```suggestion
    for (size_t i = 0; i < num_queue_items; i++) {
```